src/CanvasRenderingContext2d.cc

```diff
@@ -1266,41 +1280,74 @@ NAN_METHOD(Context2d::DrawImage) {
   // Start draw
   cairo_save(ctx);
 
-  // Scale src
-  float fx = (float) dw / sw;
-  float fy = (float) dh / sh;
+  cairo_matrix_t matrix;
+  double transforms[6];
+  cairo_get_matrix(context->context(), &matrix);
+  decompose_matrix(matrix, transforms);
+  // extract the scale value from the current transform so that we know how many pixels we
+  // need for our extra canvas in the drawImage operation.
+  double current_scale_x = transforms[1];
+  double current_scale_y = transforms[2];
```

Comment:
I think these values should just be `matrix.xx` and `matrix.yy` instead of decomposing the matrix. Try this test case -- it's broken with `transforms[1|2]` but works with `matrix.xx|yy`

```js
tests['transformed drawimage'] = function (ctx) {
  ctx.fillStyle = 'white'
  ctx.fillRect(0, 0, 200, 200)
  ctx.fillStyle = 'black'
  ctx.fillRect(5, 5, 50, 50)
  ctx.transform(1.2, 1, 1.8, 1.3, 0, 0);
  ctx.drawImage(ctx.canvas, 0, 0)
}
```

(The sign of scale_y from the decomposed matrix is negative when it should be positive.)

Reply:
if matrix.xx is always the matrix[0] position, the number that we generally associate with scaleX, that is not enough. When you have a rotation in that position there are mixed values of scaling and angle of rotation, as well as in matrix[3] or matrix.yy.

Now the negative sign, that is a point and should be fixed, let me see if i can have some code that works with any sign ( the scale is the abs value, while the image needs a flip )

Reply:
Do you have an example test case that fails using just `xx` and `yy`? My understanding of this code (which may be wrong) is that both the canvas and the image are rotated together, so xx and yy are all you need. I tried quite a few random matrices.

Also I was unclear on the "negative when it should be positive" bit -- `yy` was like ~2.0 and `transforms[2]` was like -0.6 -- not just an `abs()` would have fixed it.

Reply:
well i may have done something wrong. i ll triple check. i use this decomposing with success elsewhere in javascript, not in c.